perlin.py

```diff
@@ -0,0 +1,151 @@
+    :param radius: radius of the city
+    :param scale: the scale to multiply to each coordinate, default is 0.005
+    :param octaves: the octaves to use when sampling, default is 3
+    :return: the scaled noise value as float in [0:1]
+    """
+    x, y = get_edge_pair_centroid(edge.getShape())
+    return get_perlin_noise(x=float(x), y=float(y), base=base, scale=scale, octaves=octaves) + (
+            1 - (distance((x, y), centre) / radius))
+
+
+def apply_network_noise(net: sumolib.net.Net, xml: ElementTree, scale: float = 0.005, octaves: int = 3):
+    """
+    Calculate and apply Perlin noise in [0:1] range for each street for population and industry
+    :param net: the SUMO network
+    :param xml: the statistics XML for the network
+    :param scale: the scale to multiply to each coordinate, default is 0.005
+    :param octaves: the octaves to use when sampling, default is 3
+    :return:
+    """
+    # Calculate and apply Perlin noise for all edges in network to population in statistics
+    print("Writing Perlin noise to population and industry")
+
+    from randomActivityGen import find_city_centre
+    centre = find_city_centre(net)
+    radius = radius_of_network(net, centre)
+
+    streets = xml.find("streets")
+    if streets is None:
+        streets = ET.SubElement(xml.getroot(), "streets")
+
+    # Some edges might already have a street, so we want to ignore those
+    known_streets = {street.attrib["edge"]: street for street in streets.findall("street")}
+
+    for edge in net.getEdges():
+        eid = edge.getID()
+        if eid not in known_streets:
+            # This edge is missing a street entry. Find population and industry for this edge
+            population = get_population_number(edge=edge, base=POPULATION_BASE, scale=scale, octaves=octaves,
+                                               centre=centre, radius=radius)
+            industry = get_population_number(edge=edge, base=INDUSTRY_BASE, scale=scale, octaves=octaves,
+                                             centre=centre, radius=radius)
+
+            ET.SubElement(streets, "street", {
+                "edge": eid,
+                "population": str(population),
+                "workPosition": str(industry)
+            })
+
+
+def radius_of_network(net: sumolib.net.Net, centre):
+    """
+    Get distance from centre to outermost node. Use this for computing radius of network.
+    FIXME: if centre is given off-center, it will return the largest distance from the centre which is not the radius
+    :return: the radius of the network
+    """
+    return np.max([distance(centre, node.getCoord()) for node in net.getNodes()])
+
+
+def distance(pos1, pos2):
+    """
+    Get hypotenuse using Phytagoras theorem
+    :return: the distance between pos1 and pos2
+    """
+    x1, y1 = pos1
+    x2, y2 = pos2
+    return np.sqrt((x2 - x1) ** 2 + (y2 - y1) ** 2)
+
+
+def display_noisemap(net: sumolib.net.Net, scale: float = 0.005, octave: int = 3):
```

Comment:
This function renders a massive image. It takes 99% of the running time and even crashed the program one time when I ran it on some randomly generated networks. Let's not call it until we make it a bit more sane.

Reply:
The high runtime makes sense, as it samples many more edges than the network contains. I agree that it shouldn't be run by default, though.

Reply:
Fixed in 1156f14